### ISSUE TITLE
twisted.conch.endpoints.SSHCommandClientEndpoint.newConnection has a `None` default for ui but breaks when this is used

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6550]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6550|
|Type|defect|
|Created|2013-06-05 11:51:54Z|
|Branch|https://github.com/twisted/twisted/tree/sshendpoint-default-ui-6550|

If the `ui` parameter to `newConnection` is given the value of `None`, as it is given when no value is passed for it at all (because of its default), connections to servers not already present in the supplied known hosts file are not possible - they fail with an obscure `'NoneType' object has no attribute 'prompt'` somewhere deep inside the implementation of the endpoint.

If `None` can't actually work for `ui`, then it shouldn't be the default.  A simple change might be to make the default an object that always answers "no" to prompts.

<details><summary>Searchable metadata</summary>

```
trac-id__6550 6550
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_sshendpoint_default_ui_6550 branches/sshendpoint-default-ui-6550
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1370433114000000 1370433114000000
changetime__1373414810000000 1373414810000000
version__None None
owner__exarkun exarkun
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [39032]) Branching to 'sshendpoint-default-ui-6550'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

Looks good. There may be an additional assert or two for justifying `_NoFile`, as it seems simply using a BytesIO would make your test pass.

Please merge once done, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Tragically:

```
>>> from io import BytesIO
>>> b = BytesIO(b"no")
>>> b.write("Should I erase your hard drive?  (yes/no)")
41L
>>> b.read()
''
>>> 
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [39095]) Merge sshendpoint-default-ui-6550 

Author: exarkun
Reviewer: therve
Fixes: [#6550](https://github.com/twisted/twisted/issues/6550)

When `None` (the default) is given for the `ui` parameter to
`twisted.conch.endpoints.SSHCommandClientEndpoint.newConnection`,
use a `ConsoleUI` which will read from `/dev/tty`.  If `/dev/tty`
is not accessible, use a non-UI object that answers no to all
prompts.

This replaces the old behavior of failing when trying to use `None`
as though it were a UI object.
